src/sentry/migrations/0193_release_semver.py

```diff
@@ -0,0 +1,84 @@
+# Generated by Django 1.11.29 on 2021-05-08 00:17
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+    # This flag is used to mark that a migration shouldn't be automatically run in
+    # production. We set this to True for operations that we think are risky and want
+    # someone from ops to run manually and monitor.
+    # General advice is that if in doubt, mark your migration as `is_dangerous`.
+    # Some things you should always mark as dangerous:
+    # - Large data migrations. Typically we want these to be run manually by ops so that
+    #   they can be monitored. Since data migrations will now hold a transaction open
+    #   this is even more important.
+    # - Adding columns to highly active tables, even ones that are NULL.
+    is_dangerous = True
+
+    # This flag is used to decide whether to run this migration in a transaction or not.
+    # By default we prefer to run in a transaction, but for migrations where you want
+    # to `CREATE INDEX CONCURRENTLY` this needs to be set to False. Typically you'll
+    # want to create an index concurrently when adding one to an existing table.
+    # You'll also usually want to set this to `False` if you're writing a data
+    # migration, since we don't want the entire migration to run in one long-running
+    # transaction.
+    atomic = False
+
+    dependencies = [
+        ("sentry", "0192_remove_fileblobowner_org_fk"),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name="release",
+            name="build_code",
+            field=models.TextField(db_index=True, null=True),
+        ),
+        migrations.AddField(
+            model_name="release",
+            name="build_number",
+            field=models.BigIntegerField(db_index=True, null=True),
+        ),
+        migrations.AddField(
+            model_name="release",
+            name="major",
+            field=models.BigIntegerField(null=True),
+        ),
+        migrations.AddField(
+            model_name="release",
+            name="minor",
+            field=models.BigIntegerField(null=True),
+        ),
+        migrations.AddField(
+            model_name="release",
+            name="patch",
+            field=models.BigIntegerField(null=True),
+        ),
```

Comment:
How do you plan to compare releases when querying ? Let's say you want the releases greater than 1.2.0. Are you  going to do that in a single SQL  condition like:
```
major > 1 OR
major = 1 AND (
     minor > 2 OR
     minor = 2 AND. patch > 0
         .....
)
```

Reply:
@fpacifici Thanks for bringing this up, I'd missed the case. After some digging I found https://www.postgresql.org/docs/current/functions-comparisons.html#ROW-WISE-COMPARISON, which should handle this nicely! Might be some effort to get this to work in Django, but worse case we just fall back to sql.

Reply:
So the whole algorithm in clickhouse would have to be like

Assuming we are comparing against 1.2.1-alpha
```
(ifNull(major, -1), ifNull(minor, -1), ifNull(patch, -1)) > (1,2,1) OR
(
    (ifNull(major, -1), ifNull(minor, -1), ifNull(patch, -1)) = (1,2,1) AND 
    prerelease IS NULL 
) OR 
(
    (ifNull(major, -1), ifNull(minor, -1), ifNull(patch, -1)) = (1,2,1) AND 
    prerelease IS NOT NULL AND
    prerelease > 'alpha'  
)  
```

Unfortunately clickhouse does not compare int/strings with NULL in tuples (it always returns false), it does on arrays, but the NULL would have the highest value not the lowest.
How does postgres work for the prerelease ?

Reply:
You can drop all of the `ifNull` calls and other checks there. Assume that for any partial semvar version, we'll fill the remaining values with 0 or empty strings.
 - 1.2 becomes (1, 2, 0, 0), prerelease: ""
 - 1.2.3.4 becomes (1, 2, 3, 4), prerelease: ""
 - 1.2.3-hello becomes (1,2,3,0), prerelease:"hello"

So I think your query simplifies down to
`(major, minor, patch) > (1,2,1)`. Maybe add `prerelease > 'alpha' or prerelease = ''`. I guess your input here would have looked like '> 1.2.1-alpha'?

We need this in postgres as well, because it also treats any comparisons with null columns in the row as null as well and so fails.

Any non-semver rows will still have nulls. This is fine, since we want them excluded:
 - myversion.hello: (null, null, null, null): prerelease: null
 - 41b456beee213646a6152a8acaac07c6c52d8952: (null, null, null, null): prerelease: null

Reply:
Simplified this further to
`(major, minor, patch, case when prerelease = '' then 1 else 0 end, prerelease) = (1,2,1, 0, 'alpha')`

Reply:
I geuss, in all this, we also need to take `build_number` into account from Armin's comment. But it still works according to these rules.

This is doable in Clickhouse as well, and it won't require weird transformations when backfilling in case we will use cdc for this.